File: 875 Koko Eating Bananas.c
int isValid(int* piles, int pilesSize, int H,int k){
    int hour=0;
    for(int i=0;i<pilesSize;i++) {
        hour += piles[i]/k+(piles[i]%k!=0);
    }
    return hour>H?0:1;
}
int minEatingSpeed(int* piles, int pilesSize, int h) {
    int max=0;
    for(int i=0; i<pilesSize; i++) {
        if(max<piles[i]) {
            max = piles[i];
        }
    }
    int min = max/h+(max%h!=0);
    while(max>min) {
        int mid=(max+min)/2;
        if(isValid(piles, pilesSize,h,mid)) {
            max = mid;
        } else {
            min = mid+1;
        }
    }
    return min;
}

File: 875 Koko Eating Bananas.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
  if (piles == null || piles.length == 0) return 0;
    int max = Integer.MIN_VALUE;
    for (int p: piles) max = Math.max(max, p);
    int st = 1, ed = max;
    while (st < ed) {
      int mid = (st + ed) / 2;
      if (isPossible(piles, h, mid)) {
          ed = mid;
      }
      else st = mid + 1;
    }
    return ed;
  }
  private boolean isPossible(int[] piles, int h, int K) {
    int time = 0;
    for (int p: piles) {
      time += p / K;
      if (p % K != 0) time++;
    }
    return time <= h;
  }
}
